File: c++/Templates and STL/Stack.cpp
//Demonstration of : Stack

#include<iostream>
#define SIZE 5

using namespace std;

class Stack
{
  //data members
  int data[SIZE];//container
  int top;//topmost element

public:
  //constructor
  Stack();

  //member functions
  void push(int); //add a val
  int pop();//remove a value
  void display();//view

};


Stack :: Stack()
{
  top = -1;//indicate empty state
}

void Stack :: push(int val)//add a value
{
  if(top< SIZE-1)
  {
    top++;//raise
    data[top] = val;//insert
  }
  else
    cout<<"\n Stack Overflow";
}


int Stack :: pop()//remove and return the topmost value
{
 int val = 0;
 if(top > -1)
 {
  val = data[top];//fetch the value
  top--;//level reduces
 }
 else
  cout<<"\n Stack Empty";
 return val;
}

void Stack :: display()
{
  int i;
  for(i = top; i >=0; i--)
    cout<<"\n"<<data[i];
}

void menu()
{
  int ch;
  int val;
  Stack s;

  do
  {
    cout<<"\n 1. push ";
    cout<<"\n 2. pop ";
    cout<<"\n 3. display ";
    cout<<"\n 4. exit ";
    cout<<"\n choice : ";
    cin >> ch;

    switch(ch)
    {
      case 1://push
        cout<<"\n Enter the value to push : ";
        cin>> val;
        s.push(val);
        break;
      case 2://pop
        val = s.pop();
        cout<<"\n "<<val<<" removed";
        break;
      case 3:
        s.display();
        break;

      default:
        cout<<"\n Wrong choice ";
        break;

      case 4:
        break;
    }//switch

  }while(ch != 4);
}//menu


int main()
{
  menu();

  return 0;
}

File: python/for Python/Stack.py
class Stack:
    #constructor
    def __init__(self):
        self.mem = [] # an empty list ( a linear data structure)

    #method to add a value at the top of the stack
    def push(self, val):
        self.mem.append(val)

    #method  to remove the topmost value from the stack
    def pop(self):
        if len(self.mem) > 0:
            return self.mem.pop()
        else:
            print("Stack Underflow")
            return 0

    #method to fetch the topmost value of the stack, no deletion
    def peek(self):
        x = len(self.mem)
        if x > 0:
            return self.mem[x-1]
        else:
            print("Stack Underflow")
            return 0

    #not a valid method for stack , but we have it for test
    def display(self):
        x = len(self.mem)
        if x > 0:
            x-=1
            while x >= 0:
                print(self.mem[x])
                x-=1
        else:
            print("Stack is empty")



def main():
    #create objects of the class
    s = Stack()
    ch = 0
    while ch != 5:
        print("1 push ")
        print("2 pop ")
        print("3 peek ")
        print("4 display ")
        print("5 exit")

        ch = int(input("enter choice : "))

        if ch == 1:
            val = int(input("enter value to push "))
            s.push(val)
        elif ch == 2:
            val = s.pop()
            print(" pop : ", val)
        elif ch == 3:
            val = s.peek()
            print(" peek : ", val)
        elif ch == 4:
            s.display()
        elif ch == 5:
            break
        else:
            print("Wrong choice")


main()
